File: examples/python/advanced/error-handling/input-validation.py
from aws_durable_execution_sdk_python import (
    DurableContext,
    durable_execution,
)

@durable_execution
def handler(event: dict, context: DurableContext) -> dict:
    """Validate input and handle errors."""
    # Validate required fields
    if not event.get("user_id"):
        return {"error": "InvalidInput", "message": "user_id is required"}
    
    if not event.get("action"):
        return {"error": "InvalidInput", "message": "action is required"}
    
    # Process valid input
    user_id = event["user_id"]
    action = event["action"]
    
    result = context.step(
        lambda _: {"user_id": user_id, "action": action, "status": "completed"},
        name="process_action"
    )
    
    return result

File: examples/java/advanced/error-handling/input-validation.java
public Map<String, Object> handleRequest(Map<String, Object> event, DurableContext ctx) {
    // Validate required fields
    if (event.get("user_id") == null || ((String) event.get("user_id")).isEmpty()) {
        return Map.of("error", "InvalidInput", "message", "user_id is required");
    }

    if (event.get("action") == null || ((String) event.get("action")).isEmpty()) {
        return Map.of("error", "InvalidInput", "message", "action is required");
    }

    // Process valid input
    var userId = (String) event.get("user_id");
    var action = (String) event.get("action");

    var result = ctx.step("process_action", Map.class,
        stepCtx -> Map.of("user_id", userId, "action", action, "status", "completed"));

    return result;
}
